take a timestamp when a console logging message happens

wjwwood opened the pull request:
Also make it available for console formatting.

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3894)](http://ci.ros2.org/job/ci_linux/3894/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1013)](http://ci.ros2.org/job/ci_linux-aarch64/1013/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3220)](http://ci.ros2.org/job/ci_osx/3220/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3990)](http://ci.ros2.org/job/ci_windows/3990/)

wjwwood commented:
Missed a change in `rclcpp`'s tests, new CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3897)](http://ci.ros2.org/job/ci_linux/3897/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1014)](http://ci.ros2.org/job/ci_linux-aarch64/1014/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3221)](http://ci.ros2.org/job/ci_osx/3221/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3991)](http://ci.ros2.org/job/ci_windows/3991/)

dhood commented:
For posterity I'd like to mention an alternative that I wondered about but discarded. Rather than modifying the output handler signature to always include an rcutils timestamp (some handlers might want a different type of time to rcutils system time, or not care at all), the signature could instead be left as is and any output handlers interested in time would have to call now() themselves. The issue is that if multiple handlers are attached they might get called sequentially, in which case you want them to all have access to the time of the original call, not the time that the handler got called.

Similarly, handlers that use sim time, for example, will want the sim time of the original call as well, not when the handler was called. Since rcutils doesn't "know" about sim time, I imagine the sim timestamp would be passed to the log call in the 'context' dictionary that we have previously spoken about log calls including one day.

dhood commented on src/logging.c:
```diff
@@ -517,6 +528,24 @@ void rcutils_logging_console_output_handler(
+    } else if (strcmp("seconds", token) == 0) {
+      if (snprintf(
+          numeric_storage, sizeof(numeric_storage),
+          "%f", timestamp / 1e9) == 0)
```
Think this should check for `< 0`

wjwwood replied:
You're right, but maybe `<= 0`? If it returns zero then it means it wrote no bytes, which also seems like an issue.

dhood replied:
agreed!

wjwwood replied:
I don't think I can test this case, since I get no feedback (other than output to stderr) when this occurs (there's no return code from the function).

wjwwood replied:
I fixed the comparisons here: 7f1a17b

wjwwood commented:
> Rather than modifying the output handler signature to always include an rcutils timestamp (some handlers might want a different type of time to rcutils system time, or not care at all), the signature could instead be left as is and any output handlers interested in time would have to call now() themselves.

That's a good point, I sort of thought about that when I was deciding whether or not have a steady time stamp too. On the one hand you really want all timestamps to agree whether you log to file, console or rosout, so that would indicate take the timestamp at the time of the log call and pass it to the output handlers. On the other, someone might want, system time, steady time, or sim time. So this pr was sort of an incremental step between where we have system time and in the future we could add more timestamps or require the output handlers to get their own various timestamps themselves.

wjwwood commented:
I'm happy to change this pr how ever you think would be best. I just opened it because I wanted to see how much time passed between logs.

I'll fix the failing tests (sort of fragile since they will break anytime the file has new lines added above a certain point, but unavoidable I think).

dhood commented:
I agree that this is in the right direction. To me the contents of the PR are fine now, it's just the missing test coverage that I think should be addressed before merging

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3921)](http://ci.ros2.org/job/ci_linux/3921/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1038)](http://ci.ros2.org/job/ci_linux-aarch64/1038/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3244)](http://ci.ros2.org/job/ci_osx/3244/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4014)](http://ci.ros2.org/job/ci_windows/4014/)

dhood commented:
The new output format tokens aren't currently tested: could you please add/extend a test to cover this? Here are the other ones that check formatting: https://github.com/ros2/rcutils/pull/53/files#diff-4783eea3d9e98e3d3b03afdb62839106R36

dirk-thomas commented on src/logging.c:
```diff
@@ -518,6 +529,24 @@ void rcutils_logging_console_output_handler(
+        goto cleanup;
+      }
+      token_expansion = numeric_storage;
+    } else if (strcmp("nanoseconds", token) == 0) {
```
I guess users might use `{seconds}.{nanoseconds}` but then get confused by the variable nature of digits of the nanoseconds part. I think an additional placeholder is desired / necessary which pads the output with leasing zeros to a fixed length of 9 digits.

wjwwood replied:
I think you're confusing the purpose of `{nanoseconds}`, it's the total timestamp in nanoseconds, not the nanoseconds part of time. `{seconds}` is floating point time. So, `{seconds}.{nanoseconds}` doesn't really make any sense. See the documentation:

https://github.com/ros2/rcutils/pull/85/files#diff-f1d39d4e841937fde043a8e49b7e6c2aR62

dirk-thomas replied:
Oh, got it now. Thanks for clarifying.

dhood replied:
that was how I also first interpreted it (as sec.nanosec), so if it's happened to a few people maybe there's a way we can rename it to avoid confusion. We could just do `time` like in ROS 1 and additionally `time_ns`? Then you'd be less likely to think `{time}.{time_ns}`, but 'time' is not as descriptive as we'd like moving forward..

wjwwood replied:
It's a fair point, however, that the padding is never considered. So on the floating point on the seconds may not be consistent from machine to machine. I'll see what I can do about that, as ideally it would be similar.

I'm actually writing the missing test right now, I so I'll think if a fixed padding makes sense in either case.

wjwwood replied:
It doesn't need to be concise I don't think, we could use `timestamp` and `timestamp_as_nanoseconds`.

wjwwood replied:
@dhood and I talked about it offline and we decided on `time` and `time_as_nanoseconds`, using `time` rather than `timestamp` to be closer to what ROS 1 does, and using `_as_nanoseconds` to be maximumly specific without worrying about verbosity too much because we expect it to not be used that much in favor of the floating point seconds version.

wjwwood commented:
I factored out the conversion of the timepoints to public functions with tests. They now print fixed width, so for the nanoseconds always 20 characters (19 digits and either a `-` if negative or a leading ` ` if positive), and the seconds is always 21 characters (same as nanoseconds, but with a decimal point `.`).

I also fixed the logging tests, only the last test case was being checked due to a logical bug in the test. I also fixed a cmake style (though lint_cmake made me indent it weird) and prevented some "error being overwritten" messages in some of the tests. Those aren't really an issue, but it's nice to search for them in the CI output periodically and these were false positives.

dhood commented on include/rcutils/time.h:
```diff
@@ -100,6 +100,84 @@ RCUTILS_WARN_UNUSED
+ * The number is always fixed width, with left padding zeros up to the maximum
+ * number of digits the timepoint can represent.
+ * Right now that is 19 digits (so 19 characters) for a signed 64-bit integer.
+ * Positive timepoint values will have a leading space, whereas negative values
```
I understand that the motivation for this is so that positive and negative outputs are aligned, but isn't that only going to be necessary in like 0.1% of use cases? I think the small use case of swapping between negative and positive timestamps and wanting them to be aligned doesn't warrant the extraneous space for 99.9% of use cases (I might be missing something though). Having the default output be `[ 123.456]` is surprising to me.

wjwwood replied:
> I understand that the motivation for this is so that positive and negative outputs are aligned, but isn't that only going to be necessary in like 0.1% of use cases?

That's right, I had reservations about it too, but given I made these changes in the first place in order to try and keep a fixed width I was placing extreme value on never having a different string length. However, the case pointed out for floats (variable mantissa) is far more likely and the case I'm optimizing for here should honestly never happen (that's why I originally made the timepoint unsigned 😛).

I can change this.

dhood replied:
+1 for having the default misaligned for positive/negative numbers, if the misalignment bothers someone they can add an option for it :smile: I know we pad in some cases in ros1 e.g. `[ERROR]` vs [ INFO]` but those are much more likely.

dhood commented on include/rcutils/time.h:
```diff
@@ -100,6 +100,84 @@ RCUTILS_WARN_UNUSED
+ *
+ * \param[in] timepoint the time to be made into a string
+ * \param[out] str the output string in which it is stored
+ * \param[in] str_size the output string in which it is stored
```
this doc line is copypasta (and below)

wjwwood replied:
Oops, 94dcf14 😄

dhood commented on src/time.c:
```diff
@@ -0,0 +1,91 @@
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(str, RCUTILS_RET_INVALID_ARGUMENT, allocator)
+  if (*timepoint >= 0) {
+    // have an explicit leading space to align with negative values which start with `-`
+    if (rcutils_snprintf(str, str_size, " %.19" PRId64, *timepoint) <= 0) {
```
If `str_size` was specified as 1, the `snprintf`'s return value will be 0 from my understanding ("The number of characters that would have been written if n had been sufficiently large, not counting the terminating null character."). I don't think we should error in that case anymore. However, there's [a test below](https://github.com/ros2/rcutils/pull/85/files#diff-885f7a53192e3dffac3511e600e2afbdR159) for `str_size=1` that expects `RCUTILS_RET_OK`, is that test passing?

wjwwood replied:
All of the tests are passing (and running 😄) from what I can tell. As well as on Windows where the `snprintf` like functions some times behave slightly differently.

I think that it will return `21` or whatever it _would_ have been, regardless of the value of `str_size`, rather than returning how many it _actually_ wrote based on truncating due to `str_size` being smaller than the result.

dhood replied:
ohhh, I see. it will successfully put (0 chars + NUL) but won't _return_ 0. good good, this clears it up, thanks

dhood commented on test/test_logging_output_timestamps.regex:
```diff
@@ -0,0 +1 @@
+'[\s,-][0-9]{10}[.][0-9]{9}' '[\s,-][0-9]{19}'\n'[\s,-][0-9]{10}[.][0-9]{9}' '[\s,-][0-9]{19}'
```
`\s` instead of ` ` will also match other whitespace characters which seems unnecessarily loose. Was there a reason for not just using a space?

wjwwood replied:
Nope, I can change that and rerun the tests.

dhood replied:
no need now I was mostly just curious if there was regex trickery going on

wjwwood commented:
Ok, I rebased and addressed the comments, waiting for re-review before CI.

wjwwood commented:
Also have to make a new rclcpp pr because I messed that up, and fix rclpy warnings, before CI.

dhood commented on src/time.c:
```diff
@@ -32,57 +32,44 @@ RCUTILS_PUBLIC
+    RCUTILS_SET_ERROR_MSG("failed to format time point into string as float seconds", allocator)
+    return RCUTILS_RET_ERROR;
   }
+  // if (*time_point >= 0) {
```
leftover :smile:

wjwwood commented:
New CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4090)](http://ci.ros2.org/job/ci_linux/4090/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1186)](http://ci.ros2.org/job/ci_linux-aarch64/1186/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3385)](http://ci.ros2.org/job/ci_osx/3385/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4150)](http://ci.ros2.org/job/ci_windows/4150/)

I didn't see the `rclpy` stuff locally, and I think it might be resolved by the rebase.

dirk-thomas commented on include/rcutils/time.h:
```diff
@@ -100,6 +100,84 @@ RCUTILS_WARN_UNUSED
+ * Negative values will have a leading `-`, so they will be one character
+ * longer than the positive values.
+ *
+ * The recommended minimum size of the input string is 32 characters, but
```
How is this "recommended minimum size" relevant in the docblock? Isn't is sufficient to document the exact max size necessary? Also why 32?

Same below.

wjwwood replied:
Because the usable block size on my machine is `16` bytes, and `32` is the smallest multiple of that in which the exact size of this string fits. Also, it gives us a little flexibility for the future, e.g. if we wanted to add `+` to the front of the time to distinguish it from `-`. Not saying we will, but if when you ask for 21 characters malloc gives you 32 anyways...

This is the program I used to check:

```c++
#include <cassert>
#include <cstdio>

#if __linux__
#include <malloc.h>
#else
#include <malloc/malloc.h>
#endif

int main()
{
  uint8_t * bytes = (uint8_t *)malloc(1);
  assert(bytes);
#if __linux__
  printf("%zu\n", malloc_usable_size(bytes));
#else
  printf("%zu\n", malloc_size(bytes));
#endif
  return 0;
}
```

I didn't try it on Windows.

dhood commented:
there are failing windows tests in case you didn't notice @wjwwood

wjwwood commented:
Ok... I had to rewrite the `rcutils_snprintf()` functions, but I think I got it all fixed up.

CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4451)](http://ci.ros2.org/job/ci_linux/4451/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1386)](http://ci.ros2.org/job/ci_linux-aarch64/1386/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3635)](http://ci.ros2.org/job/ci_osx/3635/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4519)](http://ci.ros2.org/job/ci_windows/4519/)

dhood commented on src/snprintf.c:
```diff
@@ -0,0 +1,63 @@
+rcutils_vsnprintf(char * buffer, size_t buffer_size, const char * format, va_list args)
+{
+  if (
+    !(NULL == buffer && 0 == buffer_size) &&
```
I imagine the purpose of this is to allow the use case of passing (null,0,format,..) in order to get the number of characters that would-be formatted (as is done [here](https://github.com/ros2/rcutils/pull/85/files#diff-d20fbd64fcff93291e6ef221061e976aR51)). However, I still think we should ensure that `format` is not null in that case.

On windows, https://msdn.microsoft.com/en-us/library/d3xd30zz.aspx suggests to me this use case of calculating would-be characters is not supported by `_vsnprintf_s` (it will set `errno` if any are null, and we will end up returning -1 below). It would be good for portability to instead use `_vscprintf` in that case, so that `rcutils_vsnprintf` can always be used with a null buffer, at which point we can replace checks [such as this](https://github.com/ros2/rcutils/pull/85/files#diff-d20fbd64fcff93291e6ef221061e976aR48) with a single call to `rcutils_vsnprintf`. Could be left for another day though since we already have this disparity.

wjwwood replied:
That's a good point (or a couple). I'll fix it up some and get you to look again.

dhood commented on include/rcutils/snprintf.h:
```diff
@@ -20,30 +20,41 @@ extern "C"
  *
  * \see snprintf()
+ * \see _snprintf_s()
+ * \returns -1 if there is an error, but not if there is truncation
```
`vsnprintf` can return a "negative value if an error occurred", so we should either map that to -1 or use the same wording here

wjwwood commented:
@dhood I think I addressed your feedback, can you give it another look?

More CI in the meantime (always assume success 😬):

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4455)](http://ci.ros2.org/job/ci_linux/4455/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1390)](http://ci.ros2.org/job/ci_linux-aarch64/1390/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3640)](http://ci.ros2.org/job/ci_osx/3640/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4525)](http://ci.ros2.org/job/ci_windows/4525/)

dhood commented on test/test_time.cpp:
```diff
@@ -117,3 +118,109 @@ TEST_F(TestTimeFixture, test_rcutils_steady_time_now) {
+
+  const char * test_str = "should not be touched";
+  timepoint = 100;
+  (void)memmove(buffer, test_str, std::min(sizeof(buffer), strlen(test_str)));
```
Correct me if I'm wrong, but if we're gonna go the `std::min` route, then:
If `std::min` returns `sizeof(buffer)`, shouldn't we manually null terminate `buffer`?
In the other case, I think you'll want to copy `strlen(test_str)+1` (or manually null terminate `buffer`).

wjwwood replied:
You're right, I'll refactor this and add an expect case which checks for properly null terminated strings.

dhood commented on test/test_snprintf.cpp:
```diff
@@ -0,0 +1,46 @@
+  const char * test_str = "0123456789";
+  char buffer[256];
+  int ret;
+
```
be good to have a test for the `buffer=null, size=0` case, which we know has different behaviour in the underlying function on windows

wjwwood replied:
I'll add one.

dhood commented on include/rcutils/snprintf.h:
```diff
@@ -20,30 +20,47 @@ extern "C"
  * \see snprintf()
+ * \see _snprintf_s()
+ * \returns the number of bytes that would have been written given enough space,
+ *   or a negative number (usually -1) if there is an error,
```
is it useful to provide "usually -1"? It's not obvious to me how a user can make use of that info

dhood commented:
@wjwwood I rebased and addressed the two comments if you want to take a look at those commits at some point

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4692)](http://ci.ros2.org/job/ci_linux/4692/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1579)](http://ci.ros2.org/job/ci_linux-aarch64/1579/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3893)](http://ci.ros2.org/job/ci_osx/3893/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4770)](http://ci.ros2.org/job/ci_windows/4770/)

wjwwood commented:
Changes lgtm, thanks @dhood!

wjwwood commented:
Thanks for pushing this through @dhood!

dhood commented:
nooo problem!

dhood commented:
Wiki updates:
added to [logger config demo](https://github.com/ros2/ros2/wiki/Logging-and-logger-configuration#console-output-formatting)
[logger page](https://github.com/ros2/ros2/wiki/Logging#console-output-configuration) references rcutils docs which will include this change in the bouncy release
